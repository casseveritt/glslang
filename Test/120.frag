#version 120

float lowp;
float mediump;
float highp;

float precision;

in vec4 i;
out vec4 o;

uniform sampler2D s2D;
centroid varying vec2 centTexCoord;

uniform mat4x2 m;

struct s {
    float f;
};

void main()
{
    mat2x3 m23 = mat2x3(m);

    int a;
    bool b;
    s sv = s(a);
    float[2] ia = float[2](3, i.y);
    float f1 = 1;
    float f = a;
    f = a;
    ivec3 iv3;
    vec3 v3 = iv3;
    f = f + a;
    f = a - f;
    f += a;
    f = a - f;
    v3 *= iv3;
    v3 = iv3 / 2.0f;
    v3 = 3.0 * iv3;
    v3 = 2 * v3;
    v3 = v3 - 2;
    if (f <  a ||
        a <= f ||
        f >  a ||
        f >= a ||
        a == f ||
        f != a);
    f = b ? a : f;
    f = b ? f : a;
    f = b ? a : a;
    s news = sv;
    
    i.xy + i.xyz;      // ERROR
    m * i.xyz;         // ERROR
    m + i;             // ERROR
    int aoeu = 1.0;    // ERROR
    f = b;             // ERROR
    f = a + b;         // ERROR
    f = b * a;         // ERROR
    b = a;             // ERROR
    b = b + f;         // ERROR
    f |= b;            // ERROR

    gl_FragColor = texture2D(s2D, centTexCoord);

    float flat;
    float smooth;
    float noperspective;
    float uvec2;
    float uvec3;
    float uvec4;
    //packed;     // ERROR, reserved word

    {
        mat4 m;
        vec4 v;
        bool b;
        gl_FragColor += b ? v : m;  // ERROR, types don't match around ":"
    }

    gl_FragColor.xr;    // ERROR, swizzlers not from same field space
    gl_FragColor.xyxyx.xy; // ERROR, cannot make a vec5, even temporarily
    centTexCoord.z;     // ERROR, swizzler out of range
    (a,b) = true;       // ERROR, not an l-value
}

float imageBuffer;
float uimage2DRect;

int main() {}           // ERROR
void main(int a) {}     // ERROR

const int a;            // ERROR

int foo(in float a);
int foo(out float a)    // ERROR
{
    return 3.2;         // ERROR
    foo(a);             // ERROR
}

bool gen(vec3 v)
{
    if (abs(v[0]) < 1e-4F && abs(v[1]) < 1e-4)
        return true;
}

void v1()
{
}

void v2()
{
    return v1();  // ERROR, no expression allowed, even though void
}

void atest()
{
    vec4 v = gl_TexCoord[1];
    v += gl_TexCoord[3];
}

varying vec4 gl_TexCoord[6];  // okay, assigning a size
varying vec4 gl_TexCoord[5];  // ERROR, changing size

mat2x2 m22;
mat2x3 m23;
mat2x4 m24;

mat3x2 m32;
mat3x3 m33;
mat3x4 m34;

mat4x2 m42;
mat4x3 m43;
mat4x4 m44;

void foo123()
{
    mat2 r2 = matrixCompMult(m22, m22);
    mat3 r3 = matrixCompMult(m33, m33);
    mat4 r4 = matrixCompMult(m44, m44);

    mat2x3 r23 = matrixCompMult(m23, m23);
    mat2x4 r24 = matrixCompMult(m24, m24);
    mat3x2 r32 = matrixCompMult(m32, m32);
    mat3x4 r34 = matrixCompMult(m34, m34);
    mat4x2 r42 = matrixCompMult(m42, m42);
    mat4x3 r43 = matrixCompMult(m43, m43);

    mat3x2 rfoo1 = matrixCompMult(m23, m32);  // ERROR
    mat3x4 rfoo2 = matrixCompMult(m34, m44);  // ERROR    
}
