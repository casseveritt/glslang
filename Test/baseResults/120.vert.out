120.vert
ERROR: 0:3: 'in for stage inputs' : not supported for this version or the enabled extensions 
ERROR: 0:4: 'out for stage outputs' : not supported for this version or the enabled extensions 
ERROR: 0:11: 'gl_Position' : cannot add storage, auxiliary, memory, interpolation, or precision qualifier to an existing variable 
ERROR: 0:12: '' : can only have one auxiliary qualifier (centroid, patch, and sample) 
ERROR: 0:12: '' : replicated qualifiers 
ERROR: 0:12: 'foo' : identifier not previously declared 
ERROR: 0:21: 'gl_ClipDistance' : undeclared identifier 
ERROR: 0:21: 'gl_ClipDistance' :  left of '[' is not of type array, matrix, or vector  
ERROR: 0:21: 'assign' :  l-value required (can't modify a const)
ERROR: 0:28: 'length' :  array must be declared with a size before using this method
ERROR: 0:31: 'length' : incomplete method syntax 
ERROR: 0:32: 'length' : method does not accept any arguments 
ERROR: 0:33: 'flizbit' : only the length method is supported for array 
ERROR: 0:33: '=' :  cannot convert from '7-element array of float' to 'int'
ERROR: 0:34: 'flizbit' : only the length method is supported for array 
ERROR: 0:34: 'f' : can't use function syntax on variable 
ERROR: 0:34: 'a4' : redefinition 
ERROR: 0:35: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:36: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:37: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:38: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:39: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:40: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:40: 'constructor' : array constructor needs one argument per array element 
ERROR: 0:40: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:40: '=' :  cannot convert from 'const float' to '2-element array of float'
ERROR: 0:41: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:41: 'constructor' : array constructor needs one argument per array element 
ERROR: 0:41: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:41: '=' :  cannot convert from 'const float' to '2-element array of float'
ERROR: 0:50: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:51: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:52: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:53: 'arrays of arrays' : not supported with this profile: none
ERROR: 0:56: 'out' : overloaded functions must have the same parameter storage qualifiers for argument 1
ERROR: 0:57: 'float' : overloaded functions must have the same return type 
ERROR: 0:87: 'overloadC' : no matching overloaded function found 
ERROR: 0:90: 'overloadC' : no matching overloaded function found 
ERROR: 0:95: 'overloadD' : ambiguous function signature match: multiple signatures match under implicit type conversion 
ERROR: 0:98: 'overloadB' : can't use function syntax on variable 
ERROR: 0:106: 'overloadC' : no matching overloaded function found 
ERROR: 0:107: 'overloadE' : no matching overloaded function found 
ERROR: 0:108: 'overloadE' : no matching overloaded function found 
ERROR: 0:111: 'overloadE' : no matching overloaded function found 
ERROR: 0:117: 'overloadF' : no matching overloaded function found 
ERROR: 0:121: 'gl_TexCoord array size' : must be less than gl_MaxTextureCoords (32)
ERROR: 46 compilation errors.  No code generated.


ERROR: node is still EOpNull!
0:15  Function Definition: main( (void)
0:15    Function Parameters: 
0:17    Sequence
0:17      move second child to first child (2-component vector of float)
0:17        'centTexCoord' (invariant smooth out 2-component vector of float)
0:17        'attv2' (in 2-component vector of float)
0:18      move second child to first child (4-component vector of float)
0:18        'gl_Position' (invariant gl_Position 4-component vector of float)
0:18        'attv4' (in 4-component vector of float)
0:20      move second child to first child (4-component vector of float)
0:20        'gl_ClipVertex' (gl_ClipVertex 4-component vector of float)
0:20        'attv4' (in 4-component vector of float)
0:21      move second child to first child (float)
0:21        Constant:
0:21          0.000000
0:21        Constant:
0:21          0.200000
0:25      move second child to first child (4-component vector of float)
0:25        'gl_Position' (invariant gl_Position 4-component vector of float)
0:25        direct index (4-component vector of float)
0:25          'b' (12-element array of 4-component vector of float)
0:25          Constant:
0:25            11 (const int)
0:28      Sequence
0:28        move second child to first child (int)
0:28          'a1' (int)
0:28          Constant:
0:28            1 (const int)
0:30      Sequence
0:30        move second child to first child (int)
0:30          'aa' (int)
0:30          Constant:
0:30            7 (const int)
0:31      Sequence
0:31        move second child to first child (int)
0:31          'a2' (int)
0:32      Sequence
0:32        move second child to first child (int)
0:32          'a3' (int)
0:32          Constant:
0:32            12 (const int)
0:43      move second child to first child (float)
0:43        'gl_PointSize' (invariant gl_PointSize float)
0:43        Constant:
0:43          3.800000
0:78  Function Definition: foo( (void)
0:78    Function Parameters: 
0:?     Sequence
0:83      Function Call: overloadB(f1;f1; (2-component vector of float)
0:83        'f' (float)
0:83        'f' (float)
0:84      Function Call: overloadB(f1;f1; (2-component vector of float)
0:84        'f' (float)
0:84        Constant:
0:84          2 (const int)
0:85      Function Call: overloadB(f1;f1; (2-component vector of float)
0:85        Constant:
0:85          1 (const int)
0:85        'i' (int)
0:87      Constant:
0:87        0.000000
0:88      Function Call: overloadC(i1;i1; (2-component vector of float)
0:88        Constant:
0:88          1 (const int)
0:88        'i' (int)
0:89      Function Call: overloadC(vf2;vf2; (2-component vector of float)
0:89        Constant:
0:89          1.000000
0:89          1.000000
0:89        Constant:
0:89          2.000000
0:89          2.000000
0:90      Constant:
0:90        0.000000
0:91      Function Call: overloadC(vf2;vf2; (2-component vector of float)
0:91        Constant:
0:91          1 (const int)
0:91          1 (const int)
0:91        Constant:
0:91          2.000000
0:91          2.000000
0:93      Function Call: overloadD(i1;f1; (3-component vector of float)
0:93        'i' (int)
0:93        'f' (float)
0:94      Function Call: overloadD(f1;i1; (3-component vector of float)
0:94        'f' (float)
0:94        'i' (int)
0:95      Function Call: overloadD(f1;i1; (3-component vector of float)
0:95        'i' (int)
0:95        'i' (int)
0:98      Constant:
0:98        0.000000
0:100      Constant:
0:100        0.000000
0:101      Function Call: texture2D(s21;vf2; (4-component vector of float)
0:101        's2D' (uniform sampler2D)
0:101        Constant:
0:101          0 (const int)
0:101          0 (const int)
0:102      clamp (4-component vector of float)
0:102        'attv4' (in 4-component vector of float)
0:102        Constant:
0:102          0 (const int)
0:102        Constant:
0:102          1 (const int)
0:103      clamp (4-component vector of float)
0:103        Convert float to int (4-component vector of int)
0:103          'attv4' (in 4-component vector of float)
0:103        Constant:
0:103          0 (const int)
0:103        Constant:
0:103          1 (const int)
0:106      Constant:
0:106        0.000000
0:107      Constant:
0:107        0.000000
0:108      Constant:
0:108        0.000000
0:109      Function Call: overloadE(vf2; (3-component vector of float)
0:109        Constant:
0:109          3.300000
0:109          3.300000
0:110      Function Call: overloadE(mf22; (3-component vector of float)
0:110        Constant:
0:110          0.500000
0:110          0.000000
0:110          0.000000
0:110          0.500000
0:111      Constant:
0:111        0.000000
0:112      Function Call: overloadE(vf2; (3-component vector of float)
0:112        Constant:
0:112          1 (const int)
0:112          1 (const int)
0:115      Function Call: overloadE(f1[2]; (3-component vector of float)
0:115        'b' (2-element array of float)
0:117      Constant:
0:117        0.000000
0:118      Function Call: overloadF(i1; (3-component vector of float)
0:118        Constant:
0:118          1 (const int)
0:?   Linker Objects
0:?     'i' (in 4-component vector of float)
0:?     'o' (smooth out 4-component vector of float)
0:?     'attv2' (in 2-component vector of float)
0:?     'attv4' (in 4-component vector of float)
0:?     's2D' (uniform sampler2D)
0:?     'centTexCoord' (invariant smooth out 2-component vector of float)
0:?     'initted' (uniform float)
0:?       3.400000
0:?     'concall' (const float)
0:?       0.295520
0:?     'gl_TexCoord' (smooth out 35-element array of 4-component vector of float)
0:?     'gl_TexCoord' (smooth out 35-element array of 4-component vector of float)


Linked vertex stage:



